### PR TITLE
Separate assemblies for static extensions

### LINKED ISSUE #1695: Separate assembly for statically typed API extensions
There should be a greater separation between the statically typed API and the dynamic API, so that users are more aware of which one they are using. Ideally the statically typed extensions should live in a separate assembly, but they should at least be in a separate namespace.

Currently some extensions for the statically typed API are not under the StaticPipe namespace. From preliminary search the following estimators seem to have extensions that are not under Microsoft.ML.StaticPipe:
https://github.com/dotnet/machinelearning/blob/dd46276cc257e20fe09239aeedb2f25c4bded6dd/src/Microsoft.ML.Transforms/MissingValueIndicatorTransformer.cs
https://github.com/dotnet/machinelearning/blob/dd46276cc257e20fe09239aeedb2f25c4bded6dd/src/Microsoft.ML.Data/Transforms/KeyToValue.cs
https://github.com/dotnet/machinelearning/blob/dd46276cc257e20fe09239aeedb2f25c4bded6dd/src/Microsoft.ML.Transforms/KeyToVectorMapping.cs
https://github.com/dotnet/machinelearning/blob/e74bc65524430a1e0ed06912bdb5c6c21ef8d158/src/Microsoft.ML.Transforms/Text/TextFeaturizingEstimator.cs
https://github.com/dotnet/machinelearning/blob/dd46276cc257e20fe09239aeedb2f25c4bded6dd/src/Microsoft.ML.Transforms/MissingValueReplacing.cs
https://github.com/dotnet/machinelearning/blob/dd46276cc257e20fe09239aeedb2f25c4bded6dd/src/Microsoft.ML.Data/Transforms/KeyToVector.cs

### PR DESCRIPTION
**Pull request by abgoswam**
Fixes #1695 

- Created separate assemblies to host the Static API extensions. Moved the Static API extensions to the appropriate assembly.
- Requires adding [BestFriend] attribute in several places
    
1. Microsoft.ML.StaticPipe
2. Microsoft.ML.LightGBM.StaticPipe
3. Microsoft.ML.TensorFlow.StaticPipe
4. Microsoft.ML.OnnxTransform.StaticPipe
5. Microsoft.ML.HalLearners.StaticPipe

NOTE: These assemblies are named  corresponding to nuget packages (with the Microsoft.ML prefix)

### CODE REVIEW

**Comment by abgoswam** [comment]
@Zruty0 . I have created  the assemblies (to host static API extensions) so they correspond to nuget packages having the Microsoft.ML prefix. It would be great if you could take a look.

File: src/Microsoft.ML.OnnxTransform.StaticPipe/DnnImageFeaturizerStaticExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.StaticPipe;
+using Microsoft.ML.StaticPipe.Runtime;
+using Microsoft.ML.Transforms;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.ML.OnnxTransform.StaticPipe
+{
+    public static class DnnImageFeaturizerStaticExtensions
```
Comment (sfilipi) [inline src/Microsoft.ML.OnnxTransform.StaticPipe/DnnImageFeaturizerStaticExtensions.cs]: >public [](start = 4, length = 6)

ideally would want to document, but i get this is low-pri.

**Comment by sfilipi** [comment]
>             var invalidData = TextLoader.CreateReader(Env, ctx => (

I think this one should get moved  to StaticPipe too. This is a bit trickier: it is a partial class and therefore on the same namespace as the Text Loader. 

---
Refers to: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs:37 in a94d7f4. [](commit_id = a94d7f4e2d562164fc319d9d00cafb81d4cdf7e7, deletion_comment = False)

**Comment by abgoswam** [comment]
>             var invalidData = TextLoader.CreateReader(Env, ctx => (

I am thinking of keeping the TextLoader changes as a separate PR.  As you noted, it seems a bit more involved  

This PR is for the transforms / learners

---
In reply to: [448489458](https://github.com/dotnet/machinelearning/pull/1914#issuecomment-448489458) [](ancestors = 448489458)

---
Refers to: test/Microsoft.ML.Tests/Transformers/TextFeaturizerTests.cs:37 in a94d7f4. [](commit_id = a94d7f4e2d562164fc319d9d00cafb81d4cdf7e7, deletion_comment = False)

File: src/Microsoft.ML.OnnxTransform.StaticPipe/DnnImageFeaturizerStaticExtensions.cs (hunk 1 above)
Comment (abgoswam) [reply src/Microsoft.ML.OnnxTransform.StaticPipe/DnnImageFeaturizerStaticExtensions.cs]: yeap. low-pri . 

not looking to modify documentation in this PR

---
In reply to: [242805885](https://github.com/dotnet/machinelearning/pull/1914#discussion_r242805885) [](ancestors = 242805885)

**Comment by Zruty0** [comment]
> using Microsoft.ML.Core.Data;

Please make sure all the new files have header comments

---
Refers to: src/Microsoft.ML.StaticPipe/WordEmbeddingsStaticExtensions.cs:1 in a94d7f4. [](commit_id = a94d7f4e2d562164fc319d9d00cafb81d4cdf7e7, deletion_comment = False)

File: test/Microsoft.ML.StaticPipelineTesting/Training.cs [hunk 2]
```diff
@@ -21,6 +21,7 @@
 using System.Linq;
 using Xunit;
 using Xunit.Abstractions;
+using Microsoft.ML.StaticPipe;
```
Comment (Zruty0) [inline test/Microsoft.ML.StaticPipelineTesting/Training.cs]: >using [](start = 0, length = 5)

sort usings

**Comment by Zruty0** [comment]
Looks good to me with minor comments. 

I noticed that we don't yet create NuGet packages for the new static pipe assemblies. Should we do it now (maybe even in this PR), or wait? @TomFinley do you have an opinion?

---
In reply to: [448452854](https://github.com/dotnet/machinelearning/pull/1914#issuecomment-448452854) [](ancestors = 448452854)

**Pull request merged**